### PR TITLE
Modify capture helper to use the output buffer's capture functionality.

### PR DESCRIPTION
**Pull request by camertron**
Modify capture helper to use the output buffer's capture functionality.

### Motivation / Background

I work on the View Component framework which has a long-standing compatibility issue with form helpers. The problem is exactly the one identified by https://github.com/rails/rails/pull/45731, but sadly that PR only solves half of the problem.

### Detail

A [recent change](https://github.com/rails/rails/pull/45731) designed to avoid swapping references to the view context's output buffer added a `#capture` method to the `OutputBuffer` class. The original `CaptureHelper#capture` method however still swaps the buffer via `CaptureHelper#with_output_buffer`, meaning the `@output_buffer` reference is still reassigned by form helpers, etc. This change delegates capture behavior to `@output_buffer` so a single buffer is used per request.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

Co-authored with @blakewilliams

### CODE REVIEW

**Comment by casperisfine** [comment]
cc @byroot

**Comment by aaronjensen** [comment]
@camertron this change has broken haml, slim, and apparently a library you maintain: https://github.com/primer/view_components/pull/1809

I'm confused as to why this change is still here and what it actually accomplishes. Should this have been released in 7.1? Is it possible to revert it so that the other template libraries can start working again, or are they doing something fundamentally wrong and they should be corrected?

Haml and Slim now have a rather high priority issue against them and it's not clear to me if it is actually their issue or if this change is just problematic.

**Comment by byroot** [comment]
> Haml and Slim now have a rather high priority issue against them

Would you mind sharing the links?

> I'm confused as to why this change is still here and what it actually accomplishes. 

It's a performance optimization. If really Haml and Slim can't adapt we'll figure something out, but let's first see if they can be made compatible.

**Comment by aaronjensen** [comment]
Sure, they are above as referenced issues, but here they are:

https://github.com/haml/haml/issues/1153
https://github.com/slim-template/slim/issues/943

**Comment by aaronjensen** [comment]
```haml
%h1 Some heading

= capture do
  = nil
```

This template produces:

```html
<h1>Some heading</h1>
<h1>Some heading</h1>
```

Because `capture` returns `"<h1>Some heading</h1>"`. So we have a `capture` method that somehow returns the results of previous rendering /as well/ as what was actually captured. It's hard to imagine how this can be a haml bug, but again, I'm unfamiliar with this stuff.

**Comment by casperisfine** [comment]
Ok, so trying your HAML example is a fresh app:

```haml
%h1 Heading HAML

= capture do
  = nil
```

Compiles down to:

```ruby
          def _app_views_comments_index_html_haml__1158020120384174913_36460(local_assigns, output_buffer)
            @virtual_path = "comments/index";;@output_buffer = output_buffer || ActionView::OutputBuffer.new; @output_buffer.safe_concat(("<h1>Heading HAML</h1>\n".freeze)); 
; 
; _haml_compiler1 =  capture do; 
; @output_buffer.safe_concat(("\n".freeze)); 
; end; @output_buffer.safe_concat(((::Haml::Util.escape_html_safe((_haml_compiler1))).to_s)); @output_buffer
          end
```

The ERB equivalent:

```html+erb
<h1>Heading ERB</h1>
<%= capture do %>
<%= nil %>
<% end %>
```

Compiles to:

```ruby
          def _app_views_posts_index_html_erb__4208763093835704284_33480(local_assigns, output_buffer)
            @virtual_path = "posts/index";; @output_buffer.safe_append='<h1>Heading ERB</h1>
'.freeze; @output_buffer.append=  capture do ; @output_buffer.safe_append='
'.freeze; @output_buffer.append=( nil ); @output_buffer.safe_append='
'.freeze; end 
 @output_buffer.safe_append='
'.freeze;@output_buffer
          end
```

**Comment by aaronjensen** [comment]
Thanks. These are the two examples cleaned up:

```rb
def _app_views_comments_index_html_haml__1158020120384174913_36460(local_assigns, output_buffer)
  @virtual_path = "comments/index"
  @output_buffer = output_buffer || ActionView::OutputBuffer.new
  @output_buffer.safe_concat(("<h1>Heading HAML</h1>\n".freeze))

  _haml_compiler1 =  capture do
    @output_buffer.safe_concat(("\n".freeze));
  end
  @output_buffer.safe_concat(((::Haml::Util.escape_html_safe((_haml_compiler1))).to_s))
  @output_buffer
end

def _app_views_posts_index_html_erb__4208763093835704284_33480(local_assigns, output_buffer)
  @virtual_path = "posts/index"
  @output_buffer.safe_append='<h1>Heading ERB</h1>\n'.freeze
  @output_buffer.append=  capture do
    @output_buffer.safe_append='\n'.freeze
    @output_buffer.append=( nil )
    @output_buffer.safe_append='\n'.freeze
  end
  @output_buffer.safe_append='\n'.freeze
  @output_buffer
end
```

Considering these lines of code:

```
buffer = @output_buffer.capture { value = yield(*args) }

result = case string = buffer.presence || value
when OutputBuffer
  string.to_s
when ActiveSupport::SafeBuffer
  string
when String
  ERB::Util.html_escape(string)
end
```

```

In the haml example, `buffer.presence` is `nil`, and `value` is the original output buffer (it used to be the temporary capture output buffer). The difference appears to be the fact that the erb version is using the `safe_append=` alias, which will return the right hand value, rather than what it is aliased to, whereas haml uses `safe_concat`, which returns the buffer.

**Comment by casperisfine** [comment]
Yeah, but `capture` used to be:

```ruby
      def capture(*args)
        value = nil
        buffer = with_output_buffer { value = yield(*args) }
        if (string = buffer.presence || value) && string.is_a?(String)
          ERB::Util.html_escape string
        end
      end
```

I don't quite see where the meaningful difference is here. I need to investigate how it was handled in 7.0.

**Comment by aaronjensen** [comment]
The difference is it was an entirely new buffer.


Aaron

**Comment by casperisfine** [comment]
Ok, I think https://github.com/rails/rails/pull/49612 workaround the issue properly. I only checked HAML, not Slim, but I suspect the issue is the same.

It's late so I'm not gonna merge tonight, but I'll come back to it tomorrow and backport to 7.1.

**Comment by casperisfine** [comment]
In the meantime you can probably try that fix by copy pasting the `capture` method from https://github.com/rails/rails/pull/49612 into your `ApplicationHelper`.

**Comment by camertron** [comment]
Chiming in here to add some additional context.

> I'm primarily at a loss because the author had to work around this: https://github.com/primer/view_components/pull/1809 which should have been a hint that this may be a problematic commit. I'm certainly missing something.

This is definitely a bit confusing. Here's the full explanation.

I work for GitHub and am one of the maintainers of the ViewComponent framework. I also work on the primer/view_components library, which contains a sub-framework called Primer Forms that helps build Rails forms using the [Primer design system](https://primer.style/).

As @casperisfine mentioned, the change in this PR is a performance improvement... but it's also an upstream fix to a longstanding ViewComponent issue that causes out-of-order rendering in certain scenarios. The problem is especially evident when using Rails' built-in form helpers.

For this reason, I decided to write a smaller, more compact version of ViewComponent (see the [`ActsAsComponent` module](https://github.com/primer/view_components/blob/47d4c230b181f76638e144ea195f80537cef48cb/lib/primer/forms/acts_as_component.rb)) engineered to avoid the out-of-order rendering issue. Unlike ViewComponent, this "ViewComponent lite" implementation does not inherit from `ActionView::Base` and therefore does not create additional output buffers. All output is written instead to the global buffer provided by the Rails view layer.

Unfortunately I had neglected to test primer/view_components against the change, wrongly expecting it to be compatible with `ActsAsComponent`. Since our test suite runs edge Rails, our tests started failing the day after this PR was merged. More troubling though was that github.com _also_ runs edge Rails, so the change would eventually show up in production and cause issues there as well.

I found myself in the very weird position of having to re-implement the "bug" I had specifically written around to keep the whole house of cards standing. The "longer-term fix" I'm referring to in https://github.com/primer/view_components/pull/1809 is to remove `ActsAsComponent` entirely in favor of plain 'ol ViewComponents. For the time being, I chose to restore the old behavior rather than hold everything up on a time-consuming refactor.

In other words, the change in this PR is good and desirable. It's more performant, avoids out-of-order rendering issues in ViewComponent, and is more internally consistent. Apologies that it broke HAML and Slim, but it looks like @casperisfine jumped in with a viable solution. Thanks!

**Comment by aaronjensen** [comment]
@camertron out of curiosity, does byroot's fix address the issue you had with `ActsAsComponent`?

**Comment by camertron** [comment]
@aaronjensen just tested it, and I can confirm it does fix the issue with `ActsAsComponent` 👍
